src/main/java/orionmipt/LJV.java

```diff
@@ -0,0 +1,276 @@
+        if (cz.isArray())
+            return false;
+
+        do {
+            if (context.isTreatsAsPrimitive(cz)
+                    || context.isTreatsAsPrimitive(cz.getPackage())
+            )
+                return true;
+
+            if (cz == Object.class)
+                return false;
+
+            Class<?>[] ifs = cz.getInterfaces();
+            for (int i = 0; i < ifs.length; i++)
+                if (canTreatClassAsPrimitive(context, ifs[i]))
+                    return true;
+
+            cz = cz.getSuperclass();
+        } while (cz != null);
+        return false;
+    }
+
+    boolean looksLikePrimitiveArray(Object obj, Context context) {
+        Class<?> c = obj.getClass();
+        if (c.getComponentType().isPrimitive())
+            return true;
+
+        for (int i = 0, len = Array.getLength(obj); i < len; i++)
+            if (!canTreatAsPrimitive(context, Array.get(obj, i)))
+                return false;
+        return true;
+    }
+
+    private void generateDotInternal(Context ctx, IdGenerator idGenerator, Object obj, StringBuilder out, Set<Object> visited)
+            throws IllegalArgumentException {
+        if (visited.add(new VisitedObject(obj))) {
+            if (obj == null)
+                out.append(dotName(idGenerator, obj) + "[label=\"null\"" + ", shape=plaintext];\n");
+            else {
+                Class<?> c = obj.getClass();
+                if (c.isArray()) {
+                    if (looksLikePrimitiveArray(obj, ctx))
+                        processPrimitiveArray(idGenerator, obj, out);
+                    else
+                        processObjectArray(ctx, idGenerator, obj, out, visited);
+                } else {
+                    Field[] fs = c.getDeclaredFields();
+                    if (!ctx.isIgnorePrivateFields())
+                        AccessibleObject.setAccessible(fs, true);
+
+                    if (hasPrimitiveFields(ctx, fs, obj))
+                        labelObjectWithSomePrimitiveFields(ctx, idGenerator, obj, fs, out);
+                    else
+                        labelObjectWithNoPrimitiveFields(ctx, idGenerator, obj, out);
+
+                    processFields(ctx, idGenerator, obj, fs, out, visited);
+
+                    //- If we cared, we would take the trouble to check which
+                    //- fields were accessible when we started, and carefully
+                    //- restore them.  Leaving them accessible does no real harm.
+                    // if( ! ctx.ignorePrivateFields )
+                    //     AccessibleObject.setAccessible( fs, false );
+                }
+            }
+        }
+    }
+
+    /**
+     * Write a DOT digraph specification of the graph rooted at
+     * <tt>obj</tt> to <tt>out</tt>.
+     */
+    private void generateDOT(Context ctx, IdGenerator idGenerator, Object obj, StringBuilder out) {
+        out.append("digraph Java {\n");
+        generateDotInternal(ctx, idGenerator, obj, out, new HashSet<>());
```

Comment:
We could use `new Collections.newSetFromMap(new IdentityHashMap<>())` istead of `new HashSet<>()`, and there will be no need in wrapping objects into `VisitedObject`

BUT, we already have `idGenerator`, which should already keep the information about visited objects. It seems that we don't need two variables (`idGenerator` and `visited`), only one will suffice.